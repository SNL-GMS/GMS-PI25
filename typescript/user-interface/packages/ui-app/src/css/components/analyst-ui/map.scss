@import '../../gms-fonts.scss';

.map {
  width: 100%;
  height: 100%;
  padding: 0.5rem;
}
.map__status-online-indicator {
  position: absolute;
  bottom: 0.8rem;
  right: 1rem;
  color: var(--gms-main, $gms-main);
  .map__status-online-indicator__online-mode {
    color: var(--gms-main, $gms-main);
    text-shadow: -1px 0 var(--gms-recessed, $gms-recessed), 0 1px var(--gms-recessed, $gms-recessed),
      1px 0 var(--gms-recessed, $gms-recessed), 0 -1px var(--gms-recessed, $gms-recessed);
  }
  .map__status-online-indicator__offline-mode {
    color: $gms-strong-warning;
    text-shadow: -1px 0 var(--gms-recessed, $gms-recessed), 0 1px var(--gms-recessed, $gms-recessed),
      1px 0 var(--gms-recessed, $gms-recessed), 0 -1px var(--gms-recessed, $gms-recessed);
  }
}
.map__column {
  display: flex;
  height: 100%;
  flex-direction: column;
}
.map__row {
  flex: 1 1 auto;
  display: flex;
}
.map-controls {
  flex: 0 0 auto;
  display: flex;
  align-items: center;
  padding-left: 0.5rem;
}
.map-controls__label {
  margin-bottom: 0rem !important;
  margin-left: 1rem !important;
}
.map-layer-toggles {
  flex: 0 0 auto;
  width: 130px;
  display: flex;
  flex-direction: column;
  background-color: var(--gms-prominent-background, $gms-prominent-background);
}
.map__inner-column {
  flex: 1 1 auto;
  border-left: 1px solid var(--gms-background, $gms-background);
  display: flex;
  flex-direction: column;
}
.map__rendering-wrapper-1 {
  flex: 1 1 auto;
  position: relative;
}

.map__layer-button {
  position: absolute;
  z-index: 10;
  top: 5px; //to align with cesium toolbar buttons
  left: 5px;
}

.map-time {
  font-family: $gms-mono;
}

.ian-map-gl-container {
  height: 100%;
  padding: 0;
}

.layer-panel-content {
  &__container {
    margin-top: -28px;
    margin-left: 10px;
    position: absolute;
    background-color: var(--gms-popover-background, $gms-popover-background);
  }
  &--header {
    padding: 8px 8px 8px 12px;
    border-bottom: solid 1px;
  }
  &--checkbox-list {
    margin-left: 8px;
  }
}

// Disabling Columbus view
.cesium-sceneModePicker-dropDown-icon[data-bind*='COLUMBUS_VIEW'] {
  display: none;
}

.layer-panel-content--header .bp4-heading {
  color: var(--gms-prominent, $gms-prominent) !important;
}

.menu-item-create-event {
  &__lat {
    font-family: $gms-mono;
  }
  &__lon {
    padding-left: 1px;
    font-family: $gms-mono;
  }
}
