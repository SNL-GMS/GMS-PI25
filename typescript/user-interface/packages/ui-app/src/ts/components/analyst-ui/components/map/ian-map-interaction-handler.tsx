import { selectSignalDetections, useAppSelector } from '@gms/ui-state';
import * as Cesium from 'cesium';
import * as React from 'react';
import { ScreenSpaceEvent, ScreenSpaceEventHandler } from 'resium';

import type { MapHandlerProps } from '~common-ui/components/map/types';

import {
  useIANMapEventRightClickHandler,
  useSdOnRightClickHandler,
  useStationOnRightClickHandler
} from './ian-map-hooks';
import {
  ianMapEventTooltipLabel,
  ianMapHandleRightClick,
  ianMapStationTooltipLabel,
  ianMapTooltipHandleAltClick,
  ianMapTooltipHandleMouseMove
} from './ian-map-interaction-utils';
import type { MapContextMenusCallbacks } from './map-context-menus';

/**
 * This component creates and ScreenSpaceEventHandler along with a ScreenSpaceEvent of type mousemove
 * so that when an entity on the map has been hovered over a tooltip will appear.
 */
export const useIanMapInteractionHandler = (
  mapContextMenusCb: MapContextMenusCallbacks,
  setEventId: (eventId: string) => void
) => {
  // Get entity right-click handlers
  const stationRightClickHandler = useStationOnRightClickHandler(mapContextMenusCb);
  const sdRightClickHandler = useSdOnRightClickHandler(mapContextMenusCb);
  const eventRightClickHandler = useIANMapEventRightClickHandler(setEventId, mapContextMenusCb);

  const signalDetections = useAppSelector(selectSignalDetections);

  return React.useCallback(
    function IanMapInteractionHandler({ viewer }: MapHandlerProps) {
      if (viewer) {
        // check to see if we have a station tooltip entity to work with if not we add it
        if (!viewer.entities.getById('hoverLabelEntity')) {
          viewer.entities.add(ianMapStationTooltipLabel);
        }
        // check to see if we have an event tooltip entity to work with if not we add it
        if (!viewer.entities.getById('eventLabelEntity')) {
          viewer.entities.add(ianMapEventTooltipLabel);
        }
        return (
          <ScreenSpaceEventHandler key="IMTHandlers">
            <ScreenSpaceEvent
              action={event =>
                ianMapHandleRightClick(
                  event,
                  viewer,
                  mapContextMenusCb,
                  sdRightClickHandler,
                  stationRightClickHandler,
                  eventRightClickHandler
                )
              }
              type={Cesium.ScreenSpaceEventType.RIGHT_CLICK}
            />
            <ScreenSpaceEvent
              action={event =>
                ianMapTooltipHandleAltClick(
                  event,
                  viewer,
                  mapContextMenusCb,
                  Object.values(signalDetections)
                )
              }
              type={Cesium.ScreenSpaceEventType.LEFT_DOWN}
              modifier={Cesium.KeyboardEventModifier.ALT}
            />
            <ScreenSpaceEvent
              action={async event => ianMapTooltipHandleMouseMove(event, viewer)}
              type={Cesium.ScreenSpaceEventType.MOUSE_MOVE}
            />
          </ScreenSpaceEventHandler>
        );
      }
      return null;
    },
    [
      eventRightClickHandler,
      mapContextMenusCb,
      sdRightClickHandler,
      signalDetections,
      stationRightClickHandler
    ]
  );
};
