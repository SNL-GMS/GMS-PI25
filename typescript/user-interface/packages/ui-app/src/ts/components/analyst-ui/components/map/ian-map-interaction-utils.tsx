import type { SignalDetectionTypes } from '@gms/common-model';
import {
  humanReadable,
  Logger,
  secondsToString,
  TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION,
  toSentenceCase
} from '@gms/common-util';
import * as Cesium from 'cesium';
import type { CesiumMovementEvent } from 'resium';

import { TOOLTIP_HEIGHT } from '~common-ui/components/map/constants';
import {
  monoFontStyle,
  monoFontStyleNoSize
} from '~components/data-acquisition-ui/components/soh-map/constants';

import {
  getLatLonFromCesiumMovement,
  getMousePositionFromCesiumMovement,
  getObjectFromPoint,
  isSiteOrStation
} from './ian-map-utils';
import type { MapContextMenusCallbacks } from './map-context-menus';
import type { IANMapRightClickHandler } from './types';

const displayTooltipDelayMs = 200;
const labelXValue = 15;

const logger = Logger.create('GMS_LOG_MAP', process.env.GMS_LOG_MAP);

let viewerRef: Cesium.Viewer;

let buildSiteOrStationTooltipObjectCancel: () => void;
let buildEventTooltipObjectCancel: () => void;

export const IAN_MAP_TOOL_TIP_PADDING = 11;

export const ianMapStationTooltipLabelOptions: Cesium.Entity.ConstructorOptions = {
  id: 'hoverLabelEntity',
  label: {
    show: false,
    text: 'loading',
    showBackground: true,
    font: monoFontStyle,
    horizontalOrigin: Cesium.HorizontalOrigin.LEFT,
    verticalOrigin: Cesium.VerticalOrigin.TOP,
    pixelOffset: new Cesium.Cartesian2(labelXValue, 0),
    eyeOffset: new Cesium.Cartesian3(0.0, 0.0, TOOLTIP_HEIGHT)
  }
};

export const ianMapStationTooltipLabel = new Cesium.Entity(ianMapStationTooltipLabelOptions);

export const ianMapEventTooltipLabelOptions: Cesium.Entity.ConstructorOptions = {
  id: 'eventLabelEntity',
  label: {
    show: false,
    text: 'loading',
    showBackground: true,
    font: monoFontStyle,
    horizontalOrigin: Cesium.HorizontalOrigin.LEFT,
    verticalOrigin: Cesium.VerticalOrigin.TOP,
    pixelOffset: new Cesium.Cartesian2(labelXValue, 0),
    eyeOffset: new Cesium.Cartesian3(0.0, 0.0, TOOLTIP_HEIGHT)
  }
};

export const ianMapEventTooltipLabel = new Cesium.Entity(ianMapEventTooltipLabelOptions);

/**
 * Method to set the viewer
 *
 * @param viewer Cesium viewer for the utils to use
 */
export const setViewer = (viewer: Cesium.Viewer): void => {
  viewerRef = viewer;
};

/**
 * Converts workflow status string to human readable format for event tooltip
 *
 * @param workflowStatus workflow status string
 * @returns formatted string
 */
export function formatStatusForTooltipDisplay(workflowStatus: string): string {
  return toSentenceCase(humanReadable(workflowStatus));
}

/**
 * Hides the event tooltip and then removes itself from the global listener
 *
 * @param event JS event
 */
export const clearEventTooltip = (event: KeyboardEvent): void => {
  if (event?.key === 'Escape') {
    const tooltipDataSource = viewerRef?.dataSources?.getByName('Tooltip');
    const labelEntity = tooltipDataSource[0]?.entities.getById('eventLabelEntity');
    if (labelEntity) {
      labelEntity.label.show = new Cesium.ConstantProperty(false);
      document.removeEventListener('keydown', clearEventTooltip);
      viewerRef.scene.requestRender();
    }
  }
};

/**
 * Hides the event tooltip but does not remove itself from the global listener
 *
 * @param event JS event
 */
export const clearHoverTooltip = (event: KeyboardEvent): void => {
  if (event.key === 'Escape' && viewerRef) {
    const tooltipDataSource = viewerRef.dataSources?.getByName('Tooltip');
    const labelEntity = tooltipDataSource[0].entities.getById('hoverLabelEntity');
    if (labelEntity) {
      labelEntity.label.show = new Cesium.ConstantProperty(false);
      viewerRef.scene.requestRender();
    }
  }
};

/**
 * Applies style and text to a {@link Cesium.Entity} as a tooltip.
 * Tooltip is formatted with black text and a white background.
 *
 * @param labelEntity cesium label to update
 * @param labelText text to be applied to the label
 */
export const formatEntityAsTooltip = (labelEntity: Cesium.Entity, labelText: string): void => {
  const scaleFactor = 1.4;
  const horizontalPadding = 7.15;
  const verticalPadding = 4;
  const computedStyle = getComputedStyle(document.body);
  const gmsMain = computedStyle.getPropertyValue('--gms-main');
  const gmsMainInverted = computedStyle.getPropertyValue('--gms-main-inverted');
  labelEntity.label.text = new Cesium.ConstantProperty(labelText);
  labelEntity.label.backgroundColor = new Cesium.ConstantProperty(
    Cesium.Color.fromCssColorString(gmsMain)
  );
  labelEntity.label.fillColor = new Cesium.ConstantProperty(
    Cesium.Color.fromCssColorString(gmsMainInverted)
  );
  labelEntity.label.scale = new Cesium.ConstantProperty(scaleFactor);
  labelEntity.label.font = new Cesium.ConstantProperty(`10px ${monoFontStyleNoSize}`);
  labelEntity.label.backgroundPadding = new Cesium.ConstantProperty(
    new Cesium.Cartesian2(horizontalPadding, verticalPadding)
  );
};

/**
 * Builds a site or station tooltip
 *
 * @param labelEntity cesium label to update
 * @param selectedEntity selected site or station
 * @returns a Promise that wraps the site/station
 * tooltip build and display operation and a
 * callback to cancel that operation
 */
const buildSiteOrStationTooltip = (
  labelEntity: Cesium.Entity,
  viewer: Cesium.Viewer,
  movement: CesiumMovementEvent,
  selectedEntity: Cesium.Entity
): { promise: Promise<unknown>; cancel: () => void } => {
  let timer;
  const promise = new Promise<void>(resolve => {
    timer = setTimeout(() => {
      let position = null;
      let showToolTip = false;
      try {
        const point = new Cesium.Cartesian2(movement.endPosition.x, movement.endPosition.y);
        // force a pick with the globe ellipsoid to work around cesium issue
        // https://community.cesium.com/t/scene-pickposition-returns-undefined/15308/2
        position = viewer.camera.pickEllipsoid(point, viewer.scene.globe.ellipsoid);
      } catch (err) {
        logger.error(err);
      }
      if (position) {
        labelEntity.position = new Cesium.ConstantPositionProperty(position);
        formatEntityAsTooltip(labelEntity, selectedEntity.name);
        showToolTip = true;
      }
      labelEntity.label.show = new Cesium.ConstantProperty(showToolTip);
      resolve();
    }, displayTooltipDelayMs);
  });

  return {
    promise,
    cancel: () => {
      clearTimeout(timer);
    }
  };
};

/**
 * Build the hover tool tip for a signal detection
 *
 * @param labelEntity cesium label to update
 * @param viewer cesium viewer required to calculate the tooltip position
 * @param movement movement event that triggered the tooltip
 * @param properties signal detection properties
 */
const buildSignalDetectionTooltip = (
  labelEntity: Cesium.Entity,
  viewer: Cesium.Viewer,
  movement: CesiumMovementEvent,
  properties: Cesium.PropertyBag
): void => {
  let position = null;
  let showToolTip = false;
  if (!properties.phaseValue.getValue()) {
    logger.error('No phase value for signal detection!');
    labelEntity.label.show = new Cesium.ConstantProperty(false);
    return;
  }

  if (!properties.stationName.getValue()) {
    logger.error('No station name for signal detection!');
    labelEntity.label.show = new Cesium.ConstantProperty(false);
    return;
  }
  try {
    const point = new Cesium.Cartesian2(movement.endPosition.x, movement.endPosition.y);
    // force a pick with the globe ellipsoid to work around cesium issue
    // https://community.cesium.com/t/scene-pickposition-returns-undefined/15308/2
    position = viewer.camera.pickEllipsoid(point, viewer.scene.globe.ellipsoid);
  } catch (err) {
    logger.error(err);
  }
  if (position) {
    labelEntity.position = new Cesium.ConstantPositionProperty(position);
    const labelText = `${
      properties.phaseValue.getValue().value
    }-${properties.stationName.getValue()}`;
    labelEntity.label.text = new Cesium.ConstantProperty(labelText);
    showToolTip = true;
  }
  labelEntity.label.show = new Cesium.ConstantProperty(showToolTip);
};

/**
 * Build the hover tooltip for an event
 *
 * @param labelEntity cesium label to update
 * @param viewer cesium viewer required to calculate the tooltip position
 * @param movement movement event that triggered the tooltip
 * @param properties signal detection properties
 * @returns a Promise that wraps the event tooltip build/display
 * operation and a callback to cancel that operation.
 */
const buildEventTooltip = (
  labelEntity: Cesium.Entity,
  viewer: Cesium.Viewer,
  movement: CesiumMovementEvent,
  properties: Cesium.PropertyBag
) => {
  let timer;
  const promise = new Promise<void>(resolve => {
    timer = setTimeout(() => {
      let position = null;
      let showToolTip = false;
      if (!properties) {
        logger.error('No event');
        labelEntity.label.show = new Cesium.ConstantProperty(false);
        return;
      }
      try {
        const point = new Cesium.Cartesian2(movement.endPosition.x, movement.endPosition.y);
        position = viewer.camera.pickEllipsoid(point, viewer.scene.globe.ellipsoid);
      } catch (err) {
        logger.error(err);
      }
      if (position) {
        labelEntity.position = new Cesium.ConstantPositionProperty(position);
        const labelText = secondsToString(
          properties.event.getValue().time.value,
          TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION
        );
        formatEntityAsTooltip(labelEntity, labelText);
        showToolTip = true;
      }
      labelEntity.label.show = new Cesium.ConstantProperty(showToolTip);
      resolve();
    }, displayTooltipDelayMs);
  });

  return {
    promise,
    cancel: () => {
      clearTimeout(timer);
    }
  };
};

/**
 * Takes the map's label entity and adds information to it from whatever entity is found
 * at provided movement.endPosition for display as a tooltip
 * Hides the tooltip if no valid entity is found at movement.endPosition
 *
 * @param movement cesium movement
 * @param viewer The cesium map
 */
export const ianMapTooltipHandleMouseMove = async (
  movement: CesiumMovementEvent,
  viewer: Cesium.Viewer
): Promise<Cesium.Entity> => {
  const selectedEntity = getObjectFromPoint(viewer, movement.endPosition);
  const tooltipDatasource = viewer.dataSources?.getByName('Tooltip');

  viewerRef = viewer;

  if (!tooltipDatasource) {
    logger.warn('No tooltip datasource');
    return undefined;
  }
  const labelEntity = tooltipDatasource[0].entities.getById('hoverLabelEntity');

  if (!labelEntity) {
    logger.warn('No Label Entity!');
    return labelEntity;
  }

  // default to hide the tooltip.  This is set to true if a tooltip is generated
  labelEntity.label.show = new Cesium.ConstantProperty(false);

  if (buildSiteOrStationTooltipObjectCancel !== undefined) {
    buildSiteOrStationTooltipObjectCancel();
    buildSiteOrStationTooltipObjectCancel = undefined;
  }

  if (buildEventTooltipObjectCancel !== undefined) {
    buildEventTooltipObjectCancel();
    buildEventTooltipObjectCancel = undefined;
  }

  // if we are hovering over an entity
  if (selectedEntity?.properties) {
    const entityProperties = selectedEntity.properties;
    // and if this entity is a Channel Group or a Station
    if (isSiteOrStation(selectedEntity?.properties?.type?.getValue())) {
      const buildSiteOrStationToolTipObject = buildSiteOrStationTooltip(
        labelEntity,
        viewer,
        movement,
        selectedEntity
      );
      buildSiteOrStationTooltipObjectCancel = buildSiteOrStationToolTipObject.cancel;
      await buildSiteOrStationToolTipObject.promise;
    } else if (selectedEntity?.properties?.type?.getValue() === 'Signal detection') {
      buildSignalDetectionTooltip(labelEntity, viewer, movement, entityProperties);
    } else if (selectedEntity?.properties?.type?.getValue() === 'Event location') {
      const buildEventTooltipObject = buildEventTooltip(
        labelEntity,
        viewer,
        movement,
        entityProperties
      );
      buildEventTooltipObjectCancel = buildEventTooltipObject.cancel;
      await buildEventTooltipObject.promise;
    }
  }
  viewer.scene.requestRender();
  return labelEntity;
};

/**
 * Takes the map's event label entity and adds information to it
 * At provided movement.position when entity is alt-clicked
 * Adds an event listener to listen for closing tooltip with Esc
 *
 * @param movement cesium movement
 * @param viewer  cesium map
 */
export const ianMapTooltipHandleAltClick = (
  movement: Omit<CesiumMovementEvent, 'startPosition' | 'endPosition'>,
  viewer: Cesium.Viewer,
  mapContextMenusCb: MapContextMenusCallbacks,
  signalDetections: SignalDetectionTypes.SignalDetection[]
) => {
  const selectedEntity = getObjectFromPoint(viewer, movement.position);
  if (selectedEntity?.properties) {
    const tooltipPosition = getMousePositionFromCesiumMovement(movement);

    const event = new MouseEvent('contextmenu', {
      clientX: tooltipPosition.x,
      clientY: tooltipPosition.y
    });

    if (selectedEntity?.properties?.type?.getValue() === 'Event location') {
      const entityProperties = selectedEntity.properties.event.getValue();
      const { time, latitudeDegrees, longitudeDegrees, depthKm, id: eventId } = entityProperties;
      mapContextMenusCb.eventDetailsCb(event, {
        time,
        latitudeDegrees,
        longitudeDegrees,
        depthKm,
        eventId
      });
    }

    if (selectedEntity?.properties?.type?.getValue() === 'Signal detection') {
      // Used for the details and details is only enabled when one sd is selected thus safe to get the first
      const signalDetection = signalDetections.find(sd => sd.id === selectedEntity.id);

      mapContextMenusCb.signalDetectionDetailsCb(event, {
        signalDetection
      });
    }

    if (isSiteOrStation(selectedEntity?.properties?.type?.getValue())) {
      mapContextMenusCb.stationDetailsCb(event, {
        stationName: selectedEntity.properties.name.getValue(),
        latitude: selectedEntity.properties.coordinates.getValue().latitude,
        longitude: selectedEntity.properties.coordinates.getValue().longitude,
        elevation: selectedEntity.properties.coordinates.getValue().elevation,
        detailedType: selectedEntity?.properties?.statype?.getValue(), // Used for station details but not site details thus the null checks
        entityType: selectedEntity.properties.type.getValue()
      });
    }
  }
  return undefined;
};

/**
 * Determines what (if any) Cesium entity has been right-clicked and executes the
 * corresponding function.
 */
export const ianMapHandleRightClick = (
  movement: Omit<CesiumMovementEvent, 'startPosition' | 'endPosition'>,
  viewer: Cesium.Viewer,
  mapContextMenusCb: MapContextMenusCallbacks,
  sdRightClickHandler: IANMapRightClickHandler,
  stationRightClickHandler: IANMapRightClickHandler,
  eventRightClickHandler: IANMapRightClickHandler
) => {
  // Convert movement position to lat/lon
  const { latitude, longitude } = getLatLonFromCesiumMovement(movement, viewer);

  // Prepare context menu position
  const contextMenuPosition = getMousePositionFromCesiumMovement(movement);
  const event = new MouseEvent('contextmenu', {
    clientX: contextMenuPosition.x,
    clientY: contextMenuPosition.y
  });

  // Check if/type of entity selected
  const selectedEntity = getObjectFromPoint(viewer, movement.position);
  if (selectedEntity?.properties) {
    // Do not display context menu when over entities
    if (selectedEntity?.properties?.type?.getValue() === 'Event location') {
      eventRightClickHandler(movement, selectedEntity, latitude, longitude);
      return;
    }
    if (selectedEntity?.properties?.type?.getValue() === 'Signal detection') {
      sdRightClickHandler(movement, selectedEntity, latitude, longitude);
      return;
    }
    if (isSiteOrStation(selectedEntity?.properties?.type?.getValue())) {
      stationRightClickHandler(movement, selectedEntity, latitude, longitude);
      return;
    }
  }

  mapContextMenusCb.mapContextMenuCb(event, {
    latitude,
    longitude
  });
};
