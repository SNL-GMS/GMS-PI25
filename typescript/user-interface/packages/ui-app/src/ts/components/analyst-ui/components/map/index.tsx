import { IanDisplays, toDisplayTitle } from '@gms/common-model/lib/displays/types';
import { nonIdealStateWithSpinner } from '@gms/ui-core-components';
import React from 'react';

import type { IANMapComponentProps } from './types';

const IANMapLazy = React.lazy(async () =>
  import(/* webpackChunkName: 'ui-app-ian-map' */ './ian-map-component').then(module => ({
    default: module.IANMap
  }))
);

export function IANMap(props: IANMapComponentProps) {
  return (
    <React.Suspense
      fallback={nonIdealStateWithSpinner(toDisplayTitle(IanDisplays.MAP), 'Loading Display')}
    >
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <IANMapLazy {...props} />
    </React.Suspense>
  );
}
