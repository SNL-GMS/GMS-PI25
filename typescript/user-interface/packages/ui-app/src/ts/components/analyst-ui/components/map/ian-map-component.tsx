import { IanDisplays } from '@gms/common-model/lib/displays/types';
import {
  findPhaseFeatureMeasurement,
  getCurrentHypothesis
} from '@gms/common-model/lib/signal-detection/util';
import {
  selectPreviousValidActionTargetSignalDetectionIds,
  selectSelectedSdIds,
  useAllStations,
  useAppSelector,
  useUpdateSignalDetectionPhase
} from '@gms/ui-state';
import React from 'react';

import { CreateEventDialog } from '~analyst-ui/common/dialogs/create-event/create-event-dialog';
import { PhaseSelectorDialog } from '~analyst-ui/common/dialogs/phase-selector/phase-selector-dialog';
import { MapHotkeys } from '~analyst-ui/common/hotkey-configs/map-hotkey-configs';
import type { IANMapComponentProps } from '~analyst-ui/components/map/types';
import { BaseDisplay } from '~common-ui/components/base-display';

import { useVisibleSignalDetections } from '../waveform/waveform-hooks';
import {
  useIsMapSyncedToWaveformZoom,
  useMapNonPreferredEventData,
  useMapPreferredEventData
} from './ian-map-hooks';
import { IANMapPanel } from './ian-map-panel';

/**
 * IAN Map component. Renders a Cesium map and queries for Station Groups
 */
export function IANMapComponent(props: IANMapComponentProps) {
  const { glContainer } = props;
  const stationData = useAllStations();
  const signalDetectionQuery = useVisibleSignalDetections(useIsMapSyncedToWaveformZoom());
  const preferredEventData = useMapPreferredEventData();
  const nonPreferredEventData = useMapNonPreferredEventData();
  const selectedSdIds = useAppSelector(selectSelectedSdIds);

  // getting previous because the action target ids are removed from state
  // when the context menu closes, but we need to pass them to the phase menu
  const previousValidSDActionTargetIds = useAppSelector(
    selectPreviousValidActionTargetSignalDetectionIds
  );

  // CreateEventDialog setup
  const [createEventMenuVisibility, setCreateEventMenuVisibility] = React.useState(false);
  const [newEventLatitude, setNewEventLatitude] = React.useState<number | undefined>();
  const [newEventLongitude, setNewEventLongitude] = React.useState<number | undefined>();

  const setCreateEventCallback = React.useCallback(
    (visibility: boolean, lat: number, lon: number) => {
      setCreateEventMenuVisibility(visibility);
      setNewEventLatitude(lat);
      setNewEventLongitude(lon);
    },
    []
  );

  // PhaseSelectorDialog setup
  const [phaseMenuVisibility, setPhaseMenuVisibility] = React.useState(false);

  const selectedPhases = [];

  previousValidSDActionTargetIds?.forEach(id => {
    const signalDetection = signalDetectionQuery?.data?.find(sd => sd.id === id);
    if (signalDetection) {
      const currentHypothesis = getCurrentHypothesis(signalDetection.signalDetectionHypotheses);
      selectedPhases.push(
        findPhaseFeatureMeasurement(currentHypothesis.featureMeasurements).measurementValue.value
      );
    }
  });

  const signalDetectionPhaseUpdate = useUpdateSignalDetectionPhase();

  const phaseSelectorCallback = React.useCallback(
    (phases: string[]) => {
      signalDetectionPhaseUpdate(previousValidSDActionTargetIds ?? [], phases[0]);
    },
    [previousValidSDActionTargetIds, signalDetectionPhaseUpdate]
  );

  return (
    <BaseDisplay
      glContainer={glContainer}
      tabName={IanDisplays.MAP}
      className="ian-map-gl-container"
      data-cy="ian-map-container"
    >
      <MapHotkeys
        selectedSignalDetectionsIds={selectedSdIds}
        setPhaseMenuVisibility={setPhaseMenuVisibility}
        setCreateEventMenuVisibility={setCreateEventMenuVisibility}
      >
        <IANMapPanel
          stationsResult={stationData}
          signalDetections={signalDetectionQuery?.data}
          preferredEventsResult={preferredEventData}
          nonPreferredEventsResult={nonPreferredEventData}
          setPhaseMenuVisibilityCb={setPhaseMenuVisibility}
          setCreateEventCb={setCreateEventCallback}
        />
      </MapHotkeys>
      <PhaseSelectorDialog
        isOpen={phaseMenuVisibility}
        title="Set Phase"
        selectedPhases={selectedPhases}
        phaseSelectorCallback={phaseSelectorCallback}
        closeCallback={() => {
          setPhaseMenuVisibility(false);
        }}
      />
      {/* Placed outside of MapHotkeys to retain focus */}
      <CreateEventDialog
        isOpen={createEventMenuVisibility}
        onClose={() => {
          setCreateEventMenuVisibility(false);
          setNewEventLatitude(0);
          setNewEventLongitude(0);
        }}
        lat={newEventLatitude}
        lon={newEventLongitude}
      />
    </BaseDisplay>
  );
}

export const IANMap = React.memo(IANMapComponent);
