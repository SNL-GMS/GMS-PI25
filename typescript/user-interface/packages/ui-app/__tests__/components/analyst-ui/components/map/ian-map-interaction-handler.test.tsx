import { getStore } from '@gms/ui-state';
import { renderHook } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import { useIanMapInteractionHandler } from '~analyst-ui/components/map/ian-map-interaction-handler';

import { ianMapStationTooltipLabel } from '../../../../../src/ts/components/analyst-ui/components/map/ian-map-interaction-utils';

const store = getStore();

function TestReduxWrapper({ children }) {
  return <Provider store={store}>{children}</Provider>;
}

describe('IanMapInteractionHandler', () => {
  test('can create a mouse move handler', () => {
    const viewer: any = {
      scene: {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        pickPosition: jest.fn(endPosition => {
          'myPosition';
        })
      },
      entities: {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        getById: jest.fn(id => {
          return undefined;
        }),
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        add: jest.fn(incomingEntity => {
          return ianMapStationTooltipLabel;
        })
      }
    };
    const container = renderHook(
      () => {
        const IanMapTooltipHandler = useIanMapInteractionHandler(
          {
            eventContextMenuCb: jest.fn(),
            eventDetailsCb: jest.fn(),
            signalDetectionContextMenuCb: jest.fn(),
            signalDetectionDetailsCb: jest.fn(),
            stationContextMenuCb: jest.fn(),
            stationDetailsCb: jest.fn(),
            mapContextMenuCb: jest.fn()
          },
          jest.fn
        );
        return <IanMapTooltipHandler viewer={viewer} />;
      },
      { wrapper: TestReduxWrapper }
    );
    expect(container).toMatchSnapshot();
  });
  test('can handle an undefined viewer', () => {
    const viewer: any = undefined;
    const container = renderHook(
      () => {
        const IanMapTooltipHandler = useIanMapInteractionHandler(
          {
            eventContextMenuCb: jest.fn(),
            eventDetailsCb: jest.fn(),
            signalDetectionContextMenuCb: jest.fn(),
            signalDetectionDetailsCb: jest.fn(),
            stationContextMenuCb: jest.fn(),
            stationDetailsCb: jest.fn(),
            mapContextMenuCb: jest.fn()
          },
          jest.fn
        );
        return <IanMapTooltipHandler viewer={viewer} />;
      },
      { wrapper: TestReduxWrapper }
    );
    expect(container).toMatchSnapshot();
  });
});
