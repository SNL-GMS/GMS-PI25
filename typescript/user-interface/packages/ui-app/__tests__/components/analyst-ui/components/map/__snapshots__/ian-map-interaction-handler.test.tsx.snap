// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IanMapInteractionHandler can create a mouse move handler 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": <IanMapInteractionHandler
      viewer={
        {
          "entities": {
            "add": [MockFunction],
            "getById": [MockFunction],
          },
          "scene": {
            "pickPosition": [MockFunction],
          },
        }
      }
    />,
  },
  "unmount": [Function],
}
`;

exports[`IanMapInteractionHandler can handle an undefined viewer 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": <IanMapInteractionHandler />,
  },
  "unmount": [Function],
}
`;
